presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java

```diff
@@ -0,0 +1,141 @@
+ *  - Join
+ *      - AssignUniqueId (unique)
+ *          - probe (c1, c2)
+ *      - build
+ *
+ * In this case MarkDistinctNode (unique, c1, c2) is equivalent to MarkDistinctNode (unique),
+ * because if two rows match on `unique`, they must match on `c1` and `c2` as well.
+ *
+ * More generally, if input to MarkDistinctNode is grouped on a subset of distinct symbols and
+ * within each group another subset of distinct symbols is constant, then the latter subset can
+ * be removed from the list of distinct symbols.
+ *
+ * Pruning the list of distinct symbols reduces the CPU cost of hashing corresponding values and
+ * reduces the amount of memory needed to store all the distinct values.
+ */
+public class PruneMarkDistinctSymbols
+        implements Rule<MarkDistinctNode>
+{
+    private static final Pattern<MarkDistinctNode> PATTERN = markDistinct()
+            .matching(node -> node.getDistinctSymbols().size() > 1);
+
+    private final Metadata metadata;
+    private final SqlParser sqlParser;
+
+    public PruneMarkDistinctSymbols(Metadata metadata, SqlParser sqlParser)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+    }
+
+    @Override
+    public Pattern<MarkDistinctNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(MarkDistinctNode node, Captures captures, Context context)
+    {
+        // Temporary hack: deriving properties by recursively resolving equivalence groups is not ideal
+        // and won't work once the optimizer evolves to support multiple plan nodes per group.
+        List<LocalProperty<Symbol>> localProperties = derivePropertiesRecursively(node.getSource(), context.getLookup(), metadata, context.getSession(), context.getSymbolAllocator().getTypes(), sqlParser)
+                .getLocalProperties();
+
+        // Separate a set of trailing constants from the rest of the properties.
+        ImmutableSet.Builder<Symbol> prefixBuilder = ImmutableSet.builder();
+        ImmutableSet.Builder<Symbol> suffixBuilder = ImmutableSet.builder();
+        boolean foundNonConstant = false;
+        for (int i = localProperties.size() - 1; i >= 0; i--) {
+            LocalProperty<Symbol> property = localProperties.get(i);
+            if (!foundNonConstant && !(property instanceof ConstantProperty)) {
+                foundNonConstant = true;
+            }
+
+            if (foundNonConstant) {
+                prefixBuilder.addAll(property.getColumns());
+            }
+            else {
+                suffixBuilder.addAll(property.getColumns());
+            }
+        }
+
+        // Check if all non-trailing constant symbols are included in node.getDistinctSymbols().
+        // If so, the trailing constant symbols can be dropped from node.getDistinctSymbols().
+
+        if (!node.getDistinctSymbols().containsAll(prefixBuilder.build())) {
```

Comment:
Maybe add comment that all prefix symbols are implicitly `grouped`? This is definitely not obvious from the code and requires knowledge how local properties are represented and what they mean.